### PR TITLE
Reuse OAuth token across similar integrations.

### PR DESCRIPTION
**Pull request by anuragphadke**
## Proposed changes

Ability to reuse Google's OAuth token across different google integrations

## Types of changes

What types of changes does your code introduce to Pizzly?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/bearer/pizzly/blob/master/README.md#contributing-guide) guide
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

## Further comments

Currently, Pizzly doesn't support use of shared OAuth token across integrations that can reuse the token.
For eg:  I can authenticate once using Google OAuth and reuse the token across gmail/gdrive etc. because the base url "googleapis.com" is same across all integrations.

As part of custom integration, you can now use JSON config with an array of requests node that allow you:
- to reuse the same token
- `service` to distinguish what service is being used (gmail/drive etc.)
- specific headers (if any)

If `requests` node doesn't exist, the rest of the integrations will continue as usual.

Issue: https://github.com/Bearer/Pizzly/issues/218

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -96,5 +96,8 @@
     "singleQuote": true,
     "printWidth": 120,
     "tabWidth": 2
+  },
+  "pizzly": {
+    "uuid": "b92d24a3-ce04-42bc-a3e2-417277f71722"
```
Comment (cfabianski) [inline package.json:101]:
```suggestion
```

Comment (anuragphadke) [reply package.json]: done, removed UUID.

**Comment by anuragphadke** [comment]
@cfabianski @markmichon : any update on the above? is this PR helpful?

**Comment by markmichon** [comment]
@anuragphadke Seems useful. Haven't had a chance to test it yet though. Will hopefully get to it today or tomorrow.

**Comment by anuragphadke** [comment]
@markmichon : qq: do i need to do anything different to make this part of node? aka how do I build the above for `pizzly-node` ?

**Comment by markmichon** [comment]
Yeah both the js and node clients will likely need some updates to accommodate your implementation, which is why initially I was thinking it would be easier to take the low-fi approach of letting users add the remainder of the path (drive/v1 for example) onto their requests instead of building on new functionality to swap between base urls.

**Comment by anuragphadke** [comment]
no worries.. open to suggestions..

**Comment by markmichon** [comment]
Just to clarify what I mean, the integration would look like a normal one, but with a simplified config file:

```json
// e.g. google-unified.json
{
  "name": "Google",
  "auth": {
    "authorizationURL": "https://accounts.google.com/o/oauth2/v2/auth",
    "tokenURL": "https://www.googleapis.com/oauth2/v4/token",
    "authType": "OAUTH2",
    "tokenParams": { "grant_type": "authorization_code" },
    "authorizationParams": { "prompt": "consent", "access_type": "offline" },
    "auth": {
      "response_type": "code"
    }
  },
  "request": {
    "baseURL": "https://www.googleapis.com/",
    "headers": {
      "Accept": "application/json",
      "Authorization": "Bearer ${auth.accessToken}",
      "User-Agent": "Pizzly"
    }
  }
}
```

Then for usage, the clients don't need to change. Instead, users would modify their requests as if it were any other endpoint:

```js
// node client example

google.auth('valid-auth-id').get('/drive/v3/some-endpoint').then(data => { ... })

// adjust for diff services

google.auth('valid-auth-id').get('/analytics/v1/some-endpoint').then(data => { ... })
```
Is there a reason that wouldn't resolve your use case?


Your current proposed approach definitely works, but as you've mentioned it will need updates to both clients.

**Comment by anuragphadke** [comment]
@markmichon : thx for the above recommendation. It works beautifully with no code changes, sans the integration file..
thank you.

**Comment by anuragphadke** [comment]
@markmichon : GSheets actually uses a  different base url for API calls:
https://sheets.googleapis.com/v4/spreadsheets/{spreadsheetId}
Of the 155 API products by Google, https://developers.google.com/products, about 10% of them use different end-points.
Thoughts on what would be a good way to integrate with Pizzly?

**Comment by markmichon** [comment]
@anuragphadke Yeah google is quite inconsistent with their APIs. Perhaps taking the approach that the shopify integration uses would be best.  Note the `baseURL` line in the request config and ignore the auth parts. You could turn the whole base into a param, then pass it during setup. 

```json
{
  "name": "Shopify",
  "auth": {
    "authorizationURL": "https://${connectParams.shop}.myshopify.com/admin/oauth/authorize",
    "authType": "OAUTH2",
    "tokenURL": "https://${connectParams.shop}.myshopify.com/admin/oauth/access_token"
  },
  "request": {
    "baseURL": "https://${connectParams.shop}.myshopify.com/admin/api/",
    "headers": {
      "Accept": "application/json",
      "Authorization": "Bearer ${auth.accessToken}",
      "User-Agent": "Pizzly"
    },
    "params": {}
  }
}
```
Then during connection or pretty much anywhere pass in the params:
```js
await pizzlyClient.integration('shopify').connect({params:{shop:"<insert show name here>"}})
```

**Pull request closed without merging**